File: scripts/make_icon.py
#!/usr/bin/env python3
"""Generate build/icon.ico for the Windows installer.

Draws the Innovatia Learning mark (a graduation cap on an indigo tile)
directly as RGBA pixels and packs it as a multi-size PNG-in-ICO file.
Pure standard library, so it runs anywhere Python 3 does.
"""
import os
import struct
import zlib


def png_bytes(size, pixels):
    """Encode an RGBA pixel buffer (list of rows of (r,g,b,a)) as PNG."""
    def chunk(tag, data):
        c = struct.pack('>I', len(data)) + tag + data
        return c + struct.pack('>I', zlib.crc32(tag + data) & 0xFFFFFFFF)

    raw = b''
    for row in pixels:
        raw += b'\x00' + b''.join(struct.pack('BBBB', *px) for px in row)
    ihdr = struct.pack('>IIBBBBB', size, size, 8, 6, 0, 0, 0)
    return (b'\x89PNG\r\n\x1a\n'
            + chunk(b'IHDR', ihdr)
            + chunk(b'IDAT', zlib.compress(raw, 9))
            + chunk(b'IEND', b''))


def lerp(a, b, t):
    return tuple(int(a[i] + (b[i] - a[i]) * t) for i in range(3))


def render(size):
    """Render one icon frame at the given pixel size."""
    s = size
    top = (70, 85, 212)      # indigo
    bottom = (112, 72, 201)  # purple
    white = (255, 255, 255)
    gold = (244, 183, 26)

    cx = s / 2
    board_cy = s * 0.40      # centre of the mortarboard diamond
    board_w = s * 0.40       # half-width of diamond
    board_h = s * 0.16       # half-height of diamond
    base_cy = s * 0.56       # cap base (small trapezoid under the board)
    base_w = s * 0.16
    base_h = s * 0.10
    corner = s * 0.22        # tile corner radius

    rows = []
    for y in range(s):
        row = []
        for x in range(s):
            # rounded-square tile mask
            dx = max(corner - x, x - (s - 1 - corner), 0)
            dy = max(corner - y, y - (s - 1 - corner), 0)
            if dx * dx + dy * dy > corner * corner:
                row.append((0, 0, 0, 0))
                continue
            r, g, b = lerp(top, bottom, y / s)

            # mortarboard diamond
            if abs(x - cx) / board_w + abs(y - board_cy) / board_h <= 1.0:
                r, g, b = white
            # cap base
            elif abs(x - cx) <= base_w and board_cy <= y <= base_cy + base_h:
                shade = 0.88 if y > base_cy else 1.0
                r, g, b = tuple(int(c * shade) for c in white)
            # tassel: cord down the right side of the board + a bead
            tx = cx + board_w * 0.82
            if abs(x - tx) <= max(s * 0.012, 1) and board_cy <= y <= s * 0.62:
                r, g, b = gold
            bead_cy = s * 0.65
            if (x - tx) ** 2 + (y - bead_cy) ** 2 <= (s * 0.045) ** 2:
                r, g, b = gold

            row.append((r, g, b, 255))
        rows.append(row)
    return rows


def main():
    out_dir = os.path.join(os.path.dirname(__file__), '..', 'build')
    os.makedirs(out_dir, exist_ok=True)
    sizes = [256, 128, 64, 48, 32, 16]
    images = [(sz, png_bytes(sz, render(sz))) for sz in sizes]

    # ICO header + directory entries + PNG payloads
    ico = struct.pack('<HHH', 0, 1, len(images))
    offset = 6 + 16 * len(images)
    entries = b''
    payload = b''
    for sz, data in images:
        entries += struct.pack('<BBBBHHII',
                               sz % 256, sz % 256, 0, 0, 1, 32, len(data), offset)
        payload += data
        offset += len(data)

    path = os.path.join(out_dir, 'icon.ico')
    with open(path, 'wb') as f:
        f.write(ico + entries + payload)
    print(f'wrote {path} ({os.path.getsize(path)} bytes, sizes: {sizes})')

    # electron-builder also likes a plain 256px PNG for Linux/dev use
    png_path = os.path.join(out_dir, 'icon.png')
    with open(png_path, 'wb') as f:
        f.write(images[0][1])
    print(f'wrote {png_path}')


if __name__ == '__main__':
    main()

File: scripts/make-icon.js
#!/usr/bin/env node
/*
 * Generates the Innovatia app icon (build/icon.png + build/icon.ico) with no
 * external dependencies (uses Node's built-in zlib). Run automatically before
 * packaging so the binary icons don't need to be committed to the repository.
 */
const zlib = require('zlib');
const fs = require('fs');
const path = require('path');

const SIZE = 256;
const OUT = path.resolve(__dirname, '..', 'build');

const lerp = (a, b, t) => a + (b - a) * t;

// Point-in-5-point-star test.
function inStar(px, py, cx, cy, rOut, rIn, rot = -Math.PI / 2) {
  const pts = [];
  for (let i = 0; i < 10; i++) {
    const r = i % 2 === 0 ? rOut : rIn;
    const a = rot + (i * Math.PI) / 5;
    pts.push([cx + r * Math.cos(a), cy + r * Math.sin(a)]);
  }
  let inside = false;
  for (let i = 0, j = pts.length - 1; i < pts.length; j = i++) {
    const [xi, yi] = pts[i];
    const [xj, yj] = pts[j];
    if ((yi > py) !== (yj > py) && px < ((xj - xi) * (py - yi)) / (yj - yi) + xi) {
      inside = !inside;
    }
  }
  return inside;
}

function makePixels(size) {
  const cx = size / 2;
  const cy = size / 2;
  const corner = size * 0.22; // rounded-corner radius
  const px = Buffer.alloc(size * size * 4);
  let o = 0;
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const dx = Math.max(corner - x, x - (size - corner), 0);
      const dy = Math.max(corner - y, y - (size - corner), 0);
      if (dx * dx + dy * dy > corner * corner) {
        o += 4; // transparent
        continue;
      }
      const t = y / size; // brand gradient #1f4e78 -> #5b9bd5
      let r = Math.round(lerp(0x1f, 0x5b, t));
      let g = Math.round(lerp(0x4e, 0x9b, t));
      let b = Math.round(lerp(0x78, 0xd5, t));
      if (inStar(x, y, cx, cy, size * 0.34, size * 0.15)) {
        r = g = b = 255; // white star
      }
      px[o++] = r; px[o++] = g; px[o++] = b; px[o++] = 255;
    }
  }
  return px;
}

function pngBytes(size, pixels) {
  const chunk = (tag, data) => {
    const len = Buffer.alloc(4);
    len.writeUInt32BE(data.length);
    const tagBuf = Buffer.from(tag, 'ascii');
    const crc = Buffer.alloc(4);
    crc.writeUInt32BE(crc32(Buffer.concat([tagBuf, data])) >>> 0);
    return Buffer.concat([len, tagBuf, data, crc]);
  };
  const sig = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0);
  ihdr.writeUInt32BE(size, 4);
  ihdr[8] = 8; ihdr[9] = 6; // 8-bit RGBA
  const stride = size * 4;
  const raw = Buffer.alloc((stride + 1) * size);
  for (let y = 0; y < size; y++) {
    raw[y * (stride + 1)] = 0; // filter type 0
    pixels.copy(raw, y * (stride + 1) + 1, y * stride, (y + 1) * stride);
  }
  const idat = zlib.deflateSync(raw, { level: 9 });
  return Buffer.concat([sig, chunk('IHDR', ihdr), chunk('IDAT', idat), chunk('IEND', Buffer.alloc(0))]);
}

function downscale(pixels, src, dst) {
  const out = Buffer.alloc(dst * dst * 4);
  const scale = src / dst;
  let o = 0;
  for (let y = 0; y < dst; y++) {
    const sy = Math.floor(y * scale);
    for (let x = 0; x < dst; x++) {
      const sx = Math.floor(x * scale);
      const i = (sy * src + sx) * 4;
      pixels.copy(out, o, i, i + 4);
      o += 4;
    }
  }
  return out;
}

function icoBytes(images) {
  const n = images.length;
  const header = Buffer.alloc(6);
  header.writeUInt16LE(0, 0);
  header.writeUInt16LE(1, 2);
  header.writeUInt16LE(n, 4);
  const entries = [];
  const blobs = [];
  let offset = 6 + 16 * n;
  for (const { size, data } of images) {
    const e = Buffer.alloc(16);
    e[0] = size >= 256 ? 0 : size;
    e[1] = size >= 256 ? 0 : size;
    e.writeUInt16LE(1, 4); // color planes
    e.writeUInt16LE(32, 6); // bpp
    e.writeUInt32LE(data.length, 8);
    e.writeUInt32LE(offset, 12);
    entries.push(e);
    blobs.push(data);
    offset += data.length;
  }
  return Buffer.concat([header, ...entries, ...blobs]);
}

// CRC-32 (PNG polynomial).
const CRC_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    t[n] = c >>> 0;
  }
  return t;
})();
function crc32(buf) {
  let c = 0xffffffff;
  for (let i = 0; i < buf.length; i++) c = CRC_TABLE[(c ^ buf[i]) & 0xff] ^ (c >>> 8);
  return (c ^ 0xffffffff) >>> 0;
}

function main() {
  fs.mkdirSync(OUT, { recursive: true });
  const base = makePixels(SIZE);
  fs.writeFileSync(path.join(OUT, 'icon.png'), pngBytes(SIZE, base));
  const sizes = [256, 128, 64, 48, 32, 16];
  const images = sizes.map((s) => ({
    size: s,
    data: pngBytes(s, s === SIZE ? base : downscale(base, SIZE, s))
  }));
  fs.writeFileSync(path.join(OUT, 'icon.ico'), icoBytes(images));
  console.log('Generated build/icon.png and build/icon.ico (' + sizes.join(', ') + ')');
}

main();
